### PR TITLE
[HIVEMALL-138] `to_ordered_map` & `to_ordered_list` as a UDAF variant of `each_top_k`

### PR DESCRIPTION
**Pull request by takuti**
## What changes were proposed in this pull request?

- Support integer 3rd argument on `to_ordered_map` UDAF to create top/tail-k ordered map
- Create `to_ordered_list` UDAF

These functions can be used as an alternative of `each_top_k` UDTF. The main difference is that the former actively utilizes mapper-side aggregation.

## What type of PR is it?

Feature

## What is the Jira issue?

https://issues.apache.org/jira/browse/HIVEMALL-138

## How was this patch tested?

- Manual test on local and EMR
- Unit test

### map

```
to_ordered_map(primitive key, any value, [int size | boolean reverseOrder])  
```

```sql
with t as (
    select 10 as key, 'apple' as value
    union all
    select 3 as key, 'banana' as value
    union all
    select 4 as key, 'candy' as value
)
select
    to_ordered_map(key, value, true),   -- {10:"apple",4:"candy",3:"banana"} (reverse)
    to_ordered_map(key, value, 1),      -- {10:"apple"} (top-1)
    to_ordered_map(key, value, 2),      -- {10:"apple",4:"candy"} (top-2)
    to_ordered_map(key, value, 3),      -- {10:"apple",4:"candy",3:"banana"} (top-3)
    to_ordered_map(key, value, 100),    -- {10:"apple",4:"candy",3:"banana"} (top-100)
    to_ordered_map(key, value),         -- {3:"banana",4:"candy",10:"apple"} (natural)
    to_ordered_map(key, value, -1),     -- {3:"banana"} (tail-1)
    to_ordered_map(key, value, -2),     -- {3:"banana",4:"candy"} (tail-2)
    to_ordered_map(key, value, -3),     -- {3:"banana",4:"candy",10:"apple"} (tail-3)
    to_ordered_map(key, value, -100)    -- {3:"banana",4:"candy",10:"apple"} (tail-100)
from t
```

### list

- sort by value itself
  ```
  to_ordered_list(primitive value [, const string options])
  ```
- sort by key
  ```
  to_ordered_list(any value, primitive key [, const string options])
  ```

```sql
with t as (
    select 5 as key, 'apple' as value
    union all
    select 3 as key, 'banana' as value
    union all
    select 4 as key, 'candy' as value
    union all
    select 2 as key, 'donut' as value
    union all
    select 3 as key, 'egg' as value
)
select                                             -- expected output
    to_ordered_list(value, key, '-reverse'),       -- [apple, candy, (banana, egg | egg, banana), donut] (reverse order)
    to_ordered_list(value, key, '-k 2'),           -- [apple, candy] (top-k)
    to_ordered_list(value, key, '-k 100'),         -- [apple, candy, (banana, egg | egg, banana), dunut]
    to_ordered_list(value, key, '-k 2 -reverse'),  -- [donut, (banana | egg)] (reverse top-k = tail-k)
    to_ordered_list(value, key),                   -- [donut, (banana, egg | egg, banana), candy, apple] (natural order)
    to_ordered_list(value, key, '-k -2'),          -- [donut, (banana | egg)] (tail-k)
    to_ordered_list(value, key, '-k -100'),        -- [donut, (banana, egg | egg, banana), candy, apple]
    to_ordered_list(value, key, '-k -2 -reverse'), -- [apple, candy] (reverse tail-k = top-k)
    to_ordered_list(value, '-k 2'),                -- [egg, donut] (alphabetically)    
    to_ordered_list(key, '-k -2 -reverse'),        -- [5, 4] (top-2 keys)
    to_ordered_list(key)                           -- [2, 3, 3, 4, 5] (natural ordered keys)
from 
    t
;
```

> ["apple","candy","banana","egg","donut"]        ["apple","candy"]       ["apple","candy","banana","egg","donut"]        ["donut","egg"] ["donut","banana","egg","candy","apple"]    ["donut","egg"]  ["donut","banana","egg","candy","apple"]        ["apple","candy"]       ["egg","donut"] [5,4]   [2,3,3,4,5]

## Checklist

- [x] Did you apply source code formatter, i.e., `mvn formatter:format`, for your commit?

### CODE REVIEW

**Comment by takuti** [comment]
@myui This is PoC implementation. Could you give me feedback, esp. in terms of interface? I will update documentation after interface has been fixed.

File: core/src/main/java/hivemall/tools/map/UDAFToBoundedOrderedMap.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        protected PrimitiveObjectInspector sizeOI;
+
+        protected StructObjectInspector internalMergeOI;
+
+        protected StructField partialMapField;
+        protected StructField sizeField;
+
+        @Override
+        public ObjectInspector init(Mode mode, ObjectInspector[] argOIs) throws HiveException {
+            assert (argOIs.length == 3) : argOIs.length;
```
Comment (myui) [inline core/src/main/java/hivemall/tools/map/UDAFToBoundedOrderedMap.java]: argOIs.length might be 4

File: resources/ddl/define-udfs.td.hql [hunk 2]
```diff
@@ -174,6 +174,7 @@ create temporary function dimsum_mapper as 'hivemall.knn.similarity.DIMSUMMapper
 create temporary function train_classifier as 'hivemall.classifier.GeneralClassifierUDTF';
 create temporary function train_regressor as 'hivemall.regression.GeneralRegressorUDTF';
 create temporary function tree_export as 'hivemall.smile.tools.TreeExportUDF';
+create temporary function to_bounded_ordered_map as 'hivemall.tools.map.UDAFToBoundedOrderedMap';
```
Comment (myui) [inline resources/ddl/define-udfs.td.hql]: bounded ... ordered .. is annoying.  How about `to_top_k_map` ?

**Comment by myui** [comment]
How about `to_ordered_map` accepts `int k` for the third argument. Then, use `BoundedOrderedMapEvaluator`.

**Comment by takuti** [comment]
Do you mean something like:

- 3rd argument is boolean => current `to_ordered_map`
- 3rd argument is integer
  - positive => top-k ordered map (i.e., reverse ordered bounded map)
  - negative => tail-k ordered map (i.e., natural ordered bounded map)

?

Okay, it sounds reasonable for me.

**Comment by myui** [comment]
@takuti Yes. That's my intention.

**Comment by takuti** [comment]
Updated `to_ordered_map` as you suggested.

From here, I will implement `top_k_queue` to work around duplicated keys.

**Comment by myui** [comment]
`to_ordered_queue(NUMBER key, ANY value, const string options)`

options:
`-k` -- bounded queue
`-reverse`  -- to reverse order queue

**Comment by myui** [comment]
@takuti sorry, I'm considering to rename it to `to_ordered_list` .

**Comment by takuti** [comment]
@myui Ah, okay. But I imagine that "ordered list" indicates a little bit different function like `to_ordered_list(ANY value)`... Do you have any other ideas?

**Comment by takuti** [comment]
`to_ordered_list(PRIMITIVE key, ANY value)` is also acceptable for me, btw ;)

**Comment by myui** [comment]
@takuti `Do you have any other ideas?`

No corresponding data structure. Then, not to use Data structure name such as `aggr_top_k(cmpkey, value)::values`

**Comment by takuti** [comment]
I tested `each_top_k`, `to_ordered_map` and `to_ordered_list` on the same MovieLens 1M data. As we expected, `to_ordered_map` collects duplicated keys, and the number of ratings is 3 while we launched top-10 aggregation.

```sql
with topk as (
    select
        each_top_k(
            10, userid, rating,
            userid, movieid
        ) as (rank, rating, userid, movieid)
    from (
        select
            userid, movieid, rating
        from ratings
        cluster by userid
    ) t
)
select 
    count(1), collect_list(array(movieid, rating))
from 
    topk 
where 
    userid = 1
;
```

> 10      [[527.0,5.0],[3105.0,5.0],[1270.0,5.0],[48.0,5.0],[1035.0,5.0],[1193.0,5.0],[1287.0,5.0],[2355.0,5.0],[595.0,5.0],[2804.0,5.0]]

```sql
with topk as (
    select 
        userid, 
        to_ordered_map(rating, movieid, 10) as movies
    from
        ratings
    group by 
        userid
)
select 
    count(1), collect_list(array(movieid, rating))
from 
    topk
lateral view explode(movies) t as rating, movieid
where 
    userid = 1
;
```

> 3       [[2028,5],[1246,4],[745,3]]

```sql
with topk as (
    select 
        userid, 
        to_ordered_list(movieid, rating, '-k 10') as movies
    from
        ratings
    group by 
        userid
)
select 
    size(movies), movies
from 
    topk
where 
    userid = 1
;
```

> 10      [595,1035,3105,2355,1287,2804,1193,2028,1029,1270]

**Comment by myui** [comment]
that's bad...

**Comment by takuti** [comment]
what's bad...?

**Comment by myui** [comment]
lots of collisions in `to_ordered_map`.

**Comment by takuti** [comment]
Oh, that's expected behavior. I just showed how each top k, map and list are different. Using map is obviously a wrong choice in the MoviLens case.

**Comment by myui** [comment]
Failed also for to_ordered_map in Hive v2.3.0.

`IllegalArgumentException Size requested for unknown type: java.util.Map`

File: core/src/main/java/hivemall/tools/map/UDAFToOrderedMap.java [hunk 3]
```diff
@@ -92,4 +122,172 @@ public void reset(@SuppressWarnings("deprecation") AggregationBuffer agg)
...
+                Object[] parameters) throws HiveException {
+            assert (parameters.length == 3);
+
+            if (parameters[0] == null) {
+                return;
+            }
+
+            Object key = ObjectInspectorUtils.copyToStandardObject(parameters[0], inputKeyOI);
+            Object value = ObjectInspectorUtils.copyToStandardObject(parameters[1], inputValueOI);
+            int size = Math.abs(HiveUtils.getInt(parameters[2], sizeOI)); // size could be negative for tail-k
```
Comment (myui) [inline core/src/main/java/hivemall/tools/map/UDAFToOrderedMap.java]: parameter might be `boolean` but not considered.

File: core/src/main/java/hivemall/tools/map/UDAFToOrderedMap.java [hunk 4]
```diff
@@ -92,4 +122,172 @@ public void reset(@SuppressWarnings("deprecation") AggregationBuffer agg)
...
+
+        @Override
+        public ObjectInspector init(Mode mode, ObjectInspector[] argOIs) throws HiveException {
+            super.init(mode, argOIs);
+
+            // initialize input
+            if (mode == Mode.PARTIAL1 || mode == Mode.COMPLETE) {// from original data
+                this.inputKeyOI = HiveUtils.asPrimitiveObjectInspector(argOIs[0]);
+                this.inputValueOI = argOIs[1];
+                this.sizeOI = HiveUtils.asIntegerOI(argOIs[2]);
```
Comment (myui) [inline core/src/main/java/hivemall/tools/map/UDAFToOrderedMap.java]: parameter might be boolean for `argOIs[2]`

**Comment by myui** [comment]
Merged. Thanks!

**Pull request closed without merging**